core/src/main/java/com/linecorp/armeria/client/DnsResolverGroupBuilder.java

```diff
@@ -75,14 +85,119 @@ public DnsResolverGroupBuilder resolvedAddressTypes(
+     * This option is enabled by default.
+     *
+     * <p>If this option is disabled, the expired {@link DnsRecord} is removed from the {@link DnsCache} and a
+     * new {@link DnsQuery} will be executed when the next request to the domain is made.
+     */
+    @UnstableApi
+    public DnsResolverGroupBuilder enableAutoRefresh(boolean autoRefresh) {
+        this.autoRefresh = autoRefresh;
+        return this;
+    }
+
+    /**
+     * Sets the {@link Backoff} which is used when the {@link DnsNameResolver} fails to update the cache.
+     */
+    public DnsResolverGroupBuilder autoRefreshBackoff(Backoff refreshBackoff) {
+        autoRefreshBackoff = requireNonNull(refreshBackoff, "refreshBackoff");
+        return this;
+    }
+
+    /**
+     * Sets the {@link ToLongFunction} which determines how long the {@link DnsRecord}s for a hostname should
+     * be refreshed after the first cache. The {@link ToLongFunction} should return the refresh timeout in
+     * milliseconds. If a non-positive number is returned, the {@link DnsRecord}s are removed immediately
+     * without refresh.
+     *
+     * <p>Note that the timeout function is called whenever the cached {@link DnsRecord}s expire.
+     *
+     * <p>For example:
+     * <pre>{@code
+     * AtomicInteger refreshCounter = new AtomicInteger(MAX_NUM_REFRESH);
+     * dnsResolverGroupBuilder.autoRefreshTimeout(hostname -> {
+     *     if (hostname.endsWith("busy.domain.com")) {
+     *         return Duration.ofDays(7).toMillis(); // Automatically refresh the cached domain for 7 days.
+     *     }
+     *     if (hostname.endsWith("sporadic.domain.dom")) {
+     *         return 0; // Don't need to refresh a sporadically used domain.
+     *     }
+     *
+     *     if (hostname.endsWith("counting.domain.dom")) {
+     *         // Allow refreshing up to MAX_NUM_REFRESH times.
+     *         if (refreshCounter.getAndDecrease() > 0) {
+     *             return Long.MAX_VALUE;
+     *         } else {
+     *             refreshCounter.set(MAX_NUM_REFRESH);
+     *             return 0;
+     *         }
+     *     }
+     *     ...
+     * });
+     * }</pre>
+     */
+    @UnstableApi
+    public DnsResolverGroupBuilder autoRefreshTimeout(ToLongFunction<? super String> timeoutFunction) {
+        requireNonNull(timeoutFunction, "timeoutFunction");
+        //noinspection unchecked
+        autoRefreshTimeoutFunction = (ToLongFunction<String>) timeoutFunction;
+        return this;
+    }
+
+    /**
+     * Sets the timeout after which a refreshing {@link DnsRecord} should expire.
+     * If this option is unspecified and {@link #enableAutoRefresh(boolean)} is set to
+     * {@code true}, a cached {@link DnsRecord} is automatically refreshed until the {@link ClientFactory}
+     * is closed. {@link Duration#ZERO} disables the timeout.
+     */
+    @UnstableApi
+    public DnsResolverGroupBuilder autoRefreshTimeout(Duration timeout) {
```

Comment:
Unrelated to this PR, but say users want to migrate dns records.
If autoRefresh is disabled, the time taken will be `autoRefreshTimeout` + `dns.ttl` - is my understanding correct?

Reply:
No, it isn't. The expired DNS is removed immediately. 
https://github.com/line/armeria/blob/9bee536d7d8736ca6f6fd00bd35925d18c31ce6f/core/src/main/java/com/linecorp/armeria/client/RefreshingAddressResolver.java#L199-L202
Futuremore, users can't set `autoRefreshTimeout()` when autoRefresh is disabled.
https://github.com/line/armeria/blob/9bee536d7d8736ca6f6fd00bd35925d18c31ce6f/core/src/main/java/com/linecorp/armeria/client/DnsResolverGroupBuilder.java#L187-L190

Reply:
Oh sorry, I actually meant the other way around 😅 

If autoRefresh is disabled -> enabled

Reply:
I see. 😆 It isn't. `autoRefreshTimeout` doesn't increase the DNS cache TTL.
It just schedules a refreshing DNS query to update old DNS records.
So users will see new DNS records after DNS TTL.

Reply:
To be more specific, the actual TTL could be DNS's TTL + DNS query execution duration.

Reply:
Ah I think you're right - I got confused between the two caches. Thanks for the explanation 👍